qa/tasks/cephfs/test_forward_scrub.py

```diff
@@ -298,3 +299,208 @@ def test_backtrace_repair(self):
         backtrace = self.fs.read_backtrace(file_ino)
         self.assertEqual(['alpha', 'parent_a'],
                          [a['dname'] for a in backtrace['ancestors']])
+
+    def test_health_status_after_dentry_repair(self):
+        """
+        Test that the damage health status is cleared
+        after the damaged dentry is repaired
+        """
+        # Create a file for checks
+        self.mount_a.run_shell(["mkdir", "subdir/"])
+
+        self.mount_a.run_shell(["touch", "subdir/file_undamaged"])
+        self.mount_a.run_shell(["touch", "subdir/file_to_be_damaged"])
+
+        subdir_ino = self.mount_a.path_to_ino("subdir")
+
+        self.mount_a.umount_wait()
+        for mds_name in self.fs.get_active_names():
+            self.fs.mds_asok(["flush", "journal"], mds_name)
+
+        self.fs.fail()
+
+        # Corrupt a dentry
+        junk = "deadbeef" * 10
+        dirfrag_obj = "{0:x}.00000000".format(subdir_ino)
+        self.fs.radosm(["setomapval", dirfrag_obj, "file_to_be_damaged_head", junk])
+
+        # Start up and try to list it
+        self.fs.set_joinable()
+        self.fs.wait_for_daemons()
+
+        self.mount_a.mount_wait()
+        dentries = self.mount_a.ls("subdir/")
+
+        # The damaged guy should have disappeared
+        self.assertEqual(dentries, ["file_undamaged"])
+
+        # I should get ENOENT if I try and read it normally, because
+        # the dir is considered complete
+        try:
+            self.mount_a.stat("subdir/file_to_be_damaged", wait=True)
+        except CommandFailedError as e:
+            self.assertEqual(e.exitstatus, errno.ENOENT)
+        else:
+            raise AssertionError("Expected ENOENT")
+
+        nfiles = self.mount_a.getfattr("./subdir", "ceph.dir.files")
+        self.assertEqual(nfiles, "2")
+
+        self.mount_a.umount_wait()
+
+        out_json = self.fs.run_scrub(["start", "/subdir", "recursive"])
+        self.assertEqual(self.fs.wait_until_scrub_complete(tag=out_json["scrub_tag"]), True)
+
+        # Check that an entry for dentry damage is created in the damage table
+        damage = json.loads(
+            self.fs.mon_manager.raw_cluster_cmd(
+                'tell', 'mds.{0}'.format(self.fs.get_active_names()[0]),
+                "damage", "ls", '--format=json-pretty'))
+        self.assertEqual(len(damage), 1)
```

Comment:
@neesingh-rh This check fails with run: https://pulpito.ceph.com/vshankar-2023-03-02_05:17:07-fs-wip-vshankar-testing-20230228.105516-testing-default-smithi/7191542/

`dirfrag ls` gives:
```
2023-03-02T06:03:40.840 INFO:teuthology.orchestra.run.smithi119.stdout:[
2023-03-02T06:03:40.840 INFO:teuthology.orchestra.run.smithi119.stdout:    {
2023-03-02T06:03:40.840 INFO:teuthology.orchestra.run.smithi119.stdout:        "damage_type": "backtrace",
2023-03-02T06:03:40.841 INFO:teuthology.orchestra.run.smithi119.stdout:        "id": 561354219,
2023-03-02T06:03:40.841 INFO:teuthology.orchestra.run.smithi119.stdout:        "ino": 1099511627777,
2023-03-02T06:03:40.841 INFO:teuthology.orchestra.run.smithi119.stdout:        "path": "/testdir/hardlink"
2023-03-02T06:03:40.841 INFO:teuthology.orchestra.run.smithi119.stdout:    },
2023-03-02T06:03:40.842 INFO:teuthology.orchestra.run.smithi119.stdout:    {
2023-03-02T06:03:40.842 INFO:teuthology.orchestra.run.smithi119.stdout:        "damage_type": "dir_frag",
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:        "id": 2832515251,
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:        "ino": 1099511627776,
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:        "frag": "*",
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:        "path": "/dir"
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:    },
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:    {
2023-03-02T06:03:40.843 INFO:teuthology.orchestra.run.smithi119.stdout:        "damage_type": "backtrace",
2023-03-02T06:03:40.844 INFO:teuthology.orchestra.run.smithi119.stdout:        "id": 3143126501,
2023-03-02T06:03:40.844 INFO:teuthology.orchestra.run.smithi119.stdout:        "ino": 1099511627776,
2023-03-02T06:03:40.844 INFO:teuthology.orchestra.run.smithi119.stdout:        "path": "/dir"
2023-03-02T06:03:40.844 INFO:teuthology.orchestra.run.smithi119.stdout:    }
2023-03-02T06:03:40.844 INFO:teuthology.orchestra.run.smithi119.stdout:]
```

PTAL

Reply:
@vshankar During debugging for above failure, i.e. the `test_health_status_after_dirfrag_repair` I manually added more inode damages in this particular test and seeing that only the `dir_frag` damage_entry for the latest/last inode damage is being removed from the damage table. But on checking the mds logs, it shows that the repair call in `CInode.cc` for each and every inode is being visited.

Reply:
> @vshankar During debugging for above failure, i.e. the `test_health_status_after_dirfrag_repair` I manually added more inode damages in this particular test and seeing that only the `dir_frag` damage_entry for the latest/last inode damage is being removed from the damage table. But on checking the mds logs, it shows that the repair call in `CInode.cc` for each and every inode is being visited.

I see that it fails at the check for `is_dirfrag_damaged()` before the call of `erase()`

Reply:
@vshankar Here's the run: https://pulpito.ceph.com/dparmar-2023-03-13_11:40:03-fs:functional-wip-vshankar-testing-20230228.105516-distro-default-smithi/

Reply:
@neesingh-rh checking back - you fixed the original issue I reported but now you are running into a different one?

Reply:
@neesingh-rh This is a new failure where the test fails expecting a cluster log with message "bad backtrace on inode". The second time scrub is run, it does not traverse `/dir_test/file`:

```
2023-03-13T12:11:32.252+0000 7fd00043e700 10  mds.0.cache.snaprealm(0x1 seq 1 0x562638154a00) get_snaps  (seq 1 cached_seq 1)
2023-03-13T12:11:32.252+0000 7fd00043e700 15 mds.0.scrubstack scrub_dirfrag skip dentry (file,head), no change since last scrub
2023-03-13T12:11:32.252+0000 7fd00043e700 20 mds.0.cache.dir(0x10000000000) get_num_head_items() = 1; fnode.fragstat.nfiles=1 fnode.fragstat.nsubdirs=0
```

Thus the backtrace cluster log warning is not generated.